src/ray/core_worker/core_worker.cc

```diff
@@ -1378,6 +1391,64 @@ Status CoreWorker::Delete(const std::vector<ObjectID> &object_ids, bool local_on
   return plasma_store_provider_->Delete(plasma_object_ids, local_only);
 }
 
+Status CoreWorker::GetLocationFromOwner(
+    const std::vector<ObjectID> &object_ids, int64_t timeout_ms,
+    std::vector<std::shared_ptr<ObjectLocation>> *results) {
+  results->resize(object_ids.size());
+  if (object_ids.empty()) {
+    return Status::OK();
+  }
+
+  auto mutex = std::make_shared<absl::Mutex>();
+  auto num_remaining = std::make_shared<size_t>(object_ids.size());
+  auto ready_promise = std::make_shared<std::promise<void>>();
+  auto location_by_id =
+      std::make_shared<absl::flat_hash_map<ObjectID, std::shared_ptr<ObjectLocation>>>();
+
+  for (const auto &object_id : object_ids) {
+    auto owner_address = GetOwnerAddress(object_id);
+    auto client = core_worker_client_pool_->GetOrConnect(owner_address);
+    rpc::GetObjectLocationsOwnerRequest request;
+    request.set_intended_worker_id(owner_address.worker_id());
+    request.set_object_id(object_id.Binary());
+    request.set_last_version(-1);
+    client->GetObjectLocationsOwner(
+        request,
+        [object_id, mutex, num_remaining, ready_promise, location_by_id](
+            const Status &status, const rpc::GetObjectLocationsOwnerReply &reply) {
+          absl::MutexLock lock(mutex.get());
+          if (status.ok()) {
+            (*location_by_id)[object_id] = CreateObjectLocation(reply);
```

Comment:
Any particular reason for using `(*location_by_id)[object_id]` instead of `location_by_id->emplace(object_id, CreateObjectLocation(reply))` here? Do we want to overwrite the location for duplicate objects here?

Unless wishing to overwrite an existing entry, I tend to prefer `.emplace()` over `operator[]` and `.insert()` due to exception safety, single-stage construction, move/copy construction often being cheaper than move/copy assignment (and general lessening of copies), etc., but I'm happy to hear some other opinions on this. 😄

Reply:
ah I guess I was used to maps upsert semantics. I think emplace suffice in this case.